### PR TITLE
Sockets - Sarah

### PR DESCRIPTION
**Pull request by sjscotton**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
What was a design challenge that you encountered on this project? | I had a hard time minimizing dependancies in my hotel class, even though I tried to keep information about other classes and methods localized to a couple methods, it still seems highly coupled with other classes. 
What was a design decision you made that changed over time over the project? | Overall my design decisions didn't change all the much over time. One of the main things that i did switch was to make Room have a method to say if it was available on a specific day or not. This was originally in hotel, but it seemed like hotels responsibilities were too much. Putting that method in room made it easier to reuse both in Hotel and Block. 
What was a concept you gained clarity on, or a learning that you'd like to share? | Definitely learned a lot more about why too many dependancies can be detrimental. Small changes in variable names had to be changed in so many places , especially throughout tests. Next time I might create more wrapper methods for tests fo they were more adaptable to changes
What is an example of a _nominal test_ that you wrote for this assignment? What makes it a nominal case? | A nominal test would be to make sure a reservation could be made for a room that didnt have any reservations close to the reservation date. This is nominal because the expected behavior is very obvious and straight forward.
What is an example of an _edge case test_ that you wrote for this assignment? What makes it an edge case? | An Edge case would be trying to make a reservation for a room that has an existing reservation that end on the same day the new reservation will start. this is an edge case because the start date is on the edge of a different behavior. If it was on day earlier, we should get a different behavior.
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I didn't write too much pseudocode, but i did do a pretty thorough outline of my project and each class ahead of time, which I found very useful to reference while working. This is the first project i've done where I have truly followed TDD, and didn't write any code until I had written a test for it.

### CODE REVIEW

File: lib/room.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
+
+module HotelSystem
+  class Room
+    attr_reader :id, :reservations, :price_per_night
+
+    def initialize(id:, price_per_night: 200, reservations: [])
+      @id = id
+      @price_per_night = price_per_night
```
Comment (dHelmgren) [inline lib/room.rb:8]: Will price per night ever be truly variable? Why not have it as a constant?

File: lib/room.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+      @reservations = reservations
+    end
+
+    def add_reservation(reservation)
+      @reservations << reservation
+    end
+
+    def available?(date)
+      reservations.each do |reservation|
+        return false if reservation.date_range.include?(date)
```
Comment (dHelmgren) [inline lib/room.rb:18]: Having this chain of calls creates tighter coupling! You can achieve the same thing by creating an `include?` method in reservation!

File: refactor.txt [hunk 3]
```diff
@@ -0,0 +1,8 @@
+Future changes
+1. Make sure each class has a id parameter(possibly randomly created for reservations so that they each have a unique id)
+2. standardize arrive_day and first_day/ depart_day and last_day variable names throughout classes, so naming is consistent
```
Comment (dHelmgren) [inline refactor.txt:3]: yeah! This is great for consistency and readability

**Comment by dHelmgren** [comment]
# Hotel
## What We're Looking For



Feature | Feedback
---     | ---
**Baseline** |
Used git regularly | Yes 
Answer comprehension questions | Yes
**Design** |
Each class is responsible for a single piece of the program | Yes
Classes are loosely coupled | Somewhat! See comments, I tried to highlight some of the trickier ways you created dependency
**Wave 1** |
List rooms | yes
Reserve a room for a given date range | yes
List reservations for a given date | yes
Calculate reservation price | yes
Invalid date range produces an error | yes
Test coverage | 100%
**Wave 2** |
View available rooms for a given date range | Yes
Reserving a room that is not available produces an error | yes
Test coverage | 100%
**Wave 3** |
Create a block of rooms | Yes
Check if a block has rooms | yes
Reserve a room from a block | yes
Test coverage |
**Fundamentals** |
Names variables, classes and modules appropriately | Yes
Understanding of variable scope - local vs instance | Yes
Can create complex logical structures utilizing variables | yes
Appropriately uses methods to break down tasks into smaller simpler tasks | yes
Understands the differences between class and instance methods | yes
Appropriately uses iterators and `Enumerable` methods | yes
Appropriately writes and utilizes classes | yes
Appropriately utilizes modules as a namespace | yes 
**Wrap Up** |
There is a refactors.txt file | yes
The file provides a roadmap to future changes | yes
**Additional Feedback** | This is a really solid Hotel!

File: spec/block_reservation_spec.rb [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    end
+  end
+  describe "calc_total_cost" do
+    before do
+      @block_reservation = HotelSystem::BlockReservation.new(
+        room: @room,
+        arrive_day: @arrive_day,
+        depart_day: @depart_day,
+        block: @new_block,
+        discount: @discount,
```
Comment (dHelmgren) [inline spec/block_reservation_spec.rb:50]: I would check with at least 2 discounted rates!
